### PR TITLE
Add JSON format for EXPLAIN with type LOGICAL and DISTRIBUTED

### PR DESCRIPTION
**Pull request by sumeethkc**
* Add JSON format for EXPLAIN with type LOGICAL and DISTRIBUTED
* Sample output in this [gist](https://gist.github.com/sumeethkc/9ae01d844ebd31fd998cf1ef3f1dde39)

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java [hunk 1]
```diff
@@ -27,13 +30,20 @@
         implements Renderer<String>
 {
     private static final JsonCodec<JsonRenderedNode> CODEC = JsonCodec.jsonCodec(JsonRenderedNode.class);
+    private static final JsonCodec<Map<PlanFragmentId, JsonPlanFragment>> PLAN_MAP_CODEC
+            = JsonCodec.mapJsonCodec(PlanFragmentId.class, JsonPlanFragment.class);
```
Comment (sumeethkc) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java:34]: @arhimondr  regarding your comment about using a list instead of map, I thought of using json map format for both EXPLAIN and EXPLAIN ANALYZE (working on its implementation)

Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java:34]: Could you please explain why does it have to be map for explain analyze?

**Review by arhimondr — COMMENTED** [review]
Generally looks good to me. One question. Also could you please squash the commits?

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java (hunk 1 above)
Comment (sumeethkc) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java:34]: I thought a map would be better if we want to query the json output for a particular key. In explain analyze, we would have keys such as `outputLayout` (which represents the output columns for that stage) and `stageStats` which will have the cpu and mem info. Then, consumers of that output would just require to iterate over the json and query particular keys. 

I didn't really think more about list vs map.

Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java:34]: That makes sense. Thanks for the explanation.

**Review by arhimondr — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java (hunk 1 above)
Comment (sumeethkc) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java:34]: @arhimondr  thank you for the reviews.

**Comment by sumeethkc** [comment]
@martint I've sent the signed CLA to cla@trino.io. Can you please add me to the list?

**Comment by sumeethkc** [comment]
@arhimondr can you please help me with merging this PR? A maintainer has to approve to run the workflows.

**Comment by ebyhr** [comment]
@cla-bot check

**Comment by arhimondr** [comment]
@ebyhr Will you be able to merge after the tests pass?

**Comment by martint** [comment]
I want to take a deeper look at the JSON we're producing. I skimmed it a while ago and noticed some things we should adjust.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/QueryExplainer.java [hunk 2]
```diff
@@ -144,10 +146,14 @@ public String getJsonPlan(Session session, Statement statement, Type planType, L
 
         switch (planType) {
             case IO:
-                Plan plan = getLogicalPlan(session, statement, parameters, warningCollector);
-                return textIoPlan(plan, metadata, typeOperators, session);
+                Plan ioPlan = getLogicalPlan(session, statement, parameters, warningCollector);
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/QueryExplainer.java:149]: Unrelated change

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java [hunk 3]
```diff
@@ -109,4 +119,32 @@ public String getDetails()
             return remoteSources;
         }
     }
+
+    public static class JsonPlanFragment
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java:123]: To do this properly, we should introduce a `DistributedPlanRepresentation` that contains `PlanFragmentRepresentation`, which in turn contains `NodeRepresentation` objects.

Then the text and json renderers should be updated to be able to render a `DistributedPlanRepresentation` in addition to `PlanRepresentation`. `formatFragment` and `formatJsonFragmentList` should move from `PlanPrinter` into the respective renderers.

**Review by martint — CHANGES_REQUESTED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java (hunk 3 above)
Comment (sumeethkc) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java]: Refactored `formatFragmentList` into JsonRenderer. Now, `DistributedPlanRepresentation` contains all `PlanFragmentRepresentation` of the corresponding `PlanFragments`. But instead of using `NodeRepresentation` to build the json, we are utilizing `PlanFragment::getJsonRepresentation` to get the json representation directly.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/QueryExplainer.java (hunk 2 above)
Comment (sumeethkc) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/QueryExplainer.java]: Removed this.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        private final String stdDevInputRows;
+        private final String outputRows;
+        private final String outputDataSize;
+
+        @JsonCreator
+        public DistributedFragmentStats(
+                @JsonProperty("totalCpuTime") Duration totalCpuTime,
+                @JsonProperty("totalScheduledTime") Duration totalScheduledTime,
+                @JsonProperty("inputRows") String inputRows,
+                @JsonProperty("inputDataSize") String inputDataSize,
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java:154]: inputSize

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedPlanRepresentation.java [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+    private final List<DistributedPlanRepresentation> children;
+    private final Optional<JsonWindowOperatorStats> windowOperatorStats;
+
+    @JsonCreator
+    public DistributedPlanRepresentation(
+            @JsonProperty("name") String name,
+            @JsonProperty("identifier") String identifier,
+            @JsonProperty("layout") List<TypedSymbol> outputs,
+            @JsonProperty("details") String details,
+            @JsonProperty("reOrderJoinStatsAndCosts") Optional<PlanNodeStatsAndCostSummary> planNodeStatsAndCostSummary,
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedPlanRepresentation.java:59]: Why `reOrderJoinStatsAndCosts`? These are generic costs and stats, so just call it "cost"

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedPlanRepresentation.java [hunk 6]
```diff
@@ -0,0 +1,564 @@
...
+            @JsonProperty("estimates") List<EstimateStats> estimateStats,
+            @JsonProperty("planStats") DistributedPlanStats distributedPlanStats,
+            @JsonProperty("children") List<DistributedPlanRepresentation> children,
+            @JsonProperty("windowOperatorStats") Optional<JsonWindowOperatorStats> windowOperatorStats)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedPlanRepresentation.java:63]: This should not be in this class. Any stats associated with window functions should be generated directly by the window operator

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedPlanRepresentation.java [hunk 7]
```diff
@@ -0,0 +1,564 @@
...
+        private final Double stdDevHashCollisions;
+        private final DataSize planNodeSpilledDataSize;
+
+        @JsonCreator
+        public DistributedPlanStats(
+                @JsonProperty("nodeCpuTime") Duration nodeCpuTime,
+                @JsonProperty("nodeCpuFraction") Double nodeCpuFraction,
+                @JsonProperty("nodeScheduledTime") Duration nodeScheduledTime,
+                @JsonProperty("nodeScheduledFraction") Double nodeScheduledFraction,
+                @JsonProperty("nodeOutputRows") Long nodeOutputRows,
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedPlanRepresentation.java:444]: `long`

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java [hunk 8]
```diff
@@ -307,6 +314,32 @@ public static String textDistributedPlan(SubPlan plan, Metadata metadata, Sessio
...
+    public static String jsonDistributedPlan(SubPlan plan, Metadata metadata, Session session)
+    {
+        TableInfoSupplier tableInfoSupplier = new TableInfoSupplier(metadata, session);
+        ValuePrinter valuePrinter = new ValuePrinter(metadata, session);
+        TypeProvider typeProvider = getTypeProvider(plan.getAllFragments());
+        List<DistributedFragmentRepresentation> distributedFragmentRepresentations = plan.getAllFragments().stream()
+                .map(planFragment -> new JsonRenderer().formatFragment(
+                        tableInfoSupplier, ImmutableMap.of(), valuePrinter, planFragment, Optional.empty(), Optional.empty(), typeProvider, false))
+                .collect(toList());
+        log.info(distributedFragmentRepresentations.toString());
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java:339]: Remove this logging

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java [hunk 9]
```diff
@@ -211,11 +213,16 @@ private String toText(boolean verbose, int level)
         return new TextRenderer(verbose, level).render(representation);
     }
 
-    private String toJson()
+    private String toLogicalJsonPlan()
     {
         return new JsonRenderer().render(representation);
     }
 
+    public Map<Integer, List<DistributedPlanRepresentation>> toDistributedPlanRepresentation(boolean verbose)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java:221]: This method doesn't belong in this class -- it's not about rendering but about converting to DistributedPlanRepresentation.

Also, it's weird that it would rely on the JsonRenderer to do so.

The abstraction should really be:

Distributed Plan -> Distributed Plan Representation -> Textual form of the rendered plan

File: core/docker/default/etc/log.properties [hunk 10]
```diff
@@ -1,2 +1,2 @@
 # Enable verbose logging from Trino
-#io.trino=DEBUG
+io.trino=DEBUG
```
Comment (martint) [inline core/docker/default/etc/log.properties:2]: Unrelated change. Please revert it.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java [hunk 11]
```diff
@@ -54,6 +83,88 @@ private JsonRenderedNode renderJson(PlanRepresentation plan, NodeRepresentation
                         .collect(toImmutableList()));
     }
 
+    public Map<Integer, List<DistributedPlanRepresentation>> render(PlanPrinter planPrinter, boolean verbose, Integer level)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java:86]: This method doesn't do any rendering, so it's actually misnamed.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java [hunk 12]
```diff
@@ -54,6 +83,88 @@ private JsonRenderedNode renderJson(PlanRepresentation plan, NodeRepresentation
...
+    {
+        PlanRepresentation plan = planPrinter.getPlanRepresentation();
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java:88]: The method only needs the PlanRepresentation, so pass it directly instead of passing the PlanPrinter.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java [hunk 13]
```diff
@@ -54,6 +83,88 @@ private JsonRenderedNode renderJson(PlanRepresentation plan, NodeRepresentation
...
+            stageStats = stageInfo.get().getStageStats();
+
+            double avgPositionsPerTask = stageInfo.get().getTasks().stream().mapToLong(task -> task.getStats().getProcessedInputPositions()).average().orElse(Double.NaN);
+            double squaredDifferences = stageInfo.get().getTasks().stream().mapToDouble(task -> Math.pow(task.getStats().getProcessedInputPositions() - avgPositionsPerTask, 2)).sum();
+            double sdAmongTasks = Math.sqrt(squaredDifferences / stageInfo.get().getTasks().size());
+
+            distributedFragmentStats = new DistributedFragmentStats(
+                    stageStats.getTotalCpuTime().convertToMostSuccinctTimeUnit(),
+                    stageStats.getTotalScheduledTime().convertToMostSuccinctTimeUnit(),
+                    formatPositions(stageStats.getProcessedInputPositions()),
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java:117]: The DistributedFragmentRepresentation should contain unformatted values. That's up to the renderer to do when going from DistributedPlanRepresention -> textual form

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java [hunk 14]
```diff
@@ -54,6 +83,88 @@ private JsonRenderedNode renderJson(PlanRepresentation plan, NodeRepresentation
...
+    public Map<Integer, List<DistributedPlanRepresentation>> render(PlanPrinter planPrinter, boolean verbose, Integer level)
+    {
+        PlanRepresentation plan = planPrinter.getPlanRepresentation();
+        NodeRepresentation node = plan.getRoot();
+        Map<Integer, List<DistributedPlanRepresentation>> distributedPlanRepresentationMap = new HashMap<>();
+        distributedPlanRepresentationMap.computeIfAbsent(level, i -> new ArrayList<>()).add(getPlanRepresentation(plan, node, verbose, level));
+        return distributedPlanRepresentationMap;
+    }
+
+    public DistributedFragmentRepresentation formatFragment(
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/JsonRenderer.java:95]: This is also misnamed. The method is about turning plan fragment into a DistributedFragmentRepresentation

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java [hunk 15]
```diff
@@ -169,8 +168,8 @@
     private final Map<DynamicFilterId, DynamicFilterDomainStats> dynamicFilterDomainStats;
     private final ValuePrinter valuePrinter;
 
-    // NOTE: do NOT add Metadata or Session to this class.  The plan printer must be usable outside of a transaction.
-    private PlanPrinter(
+    // NOTE: do NOT add Metadata or Session to this class.  The plan printer must be usable outside a transaction.
+    public PlanPrinter(
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java:172]: Keep this constructor private and remove the call to create an instance from within the JsonRenderer. The relationship should go the other way. PlanPrinter should rely on JsonRenderer to render a plan as a json string.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/WindowOperatorStatsRepresentation.java [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+{
+    private final int activeDrivers;
+    private final int totalDrivers;
+    private final double indexSizeStdDev;
+    private final double indexPositionsStdDev;
+    private final double indexCountPerDriverStdDev;
+    private final double rowsPerDriverStdDev;
+    private final double partitionRowsStdDev;
+
+    /*@JsonCreator
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/WindowOperatorStatsRepresentation.java:29]: Remove commented out code

**Comment by martint** [comment]
A few more comments on the high level structure of the abstractions:

* The Renderer interface should have an additional `T render(DistributedPlanRepresentation plan)` method
* There should be a DistributedPlanRepresentation that contains FragmentRepresentation and any other plan-wide info necessary
* FragmentRepresentation should contain a PlanRepresentation and any other fragment-level info necessary
* FragmentRepresentation should not contain DistributedPlanRepresentation, as that should be the root of the containment hierarchy.

**Comment by sumeethkc** [comment]
> A few more comments on the high level structure of the abstractions:
> 
>     * The Renderer interface should have an additional `T render(DistributedPlanRepresentation plan)` method
> 
>     * There should be a DistributedPlanRepresentation that contains FragmentRepresentation and any other plan-wide info necessary
> 
>     * FragmentRepresentation should contain a PlanRepresentation and any other fragment-level info necessary
> 
>     * FragmentRepresentation should not contain DistributedPlanRepresentation, as that should be the root of the containment hierarchy.

`DistributedPlanRepresentation` is only used by the `JsonRenderer`. Extending it to `TextRenderer` might not be applicable. Curious to know your thoughts on this.

Also, since I am using `PlanRepresentation` which is created in `PlanPrinter` to get all the `NodeRepresentation` fields, is it necessary to encapsulate it inside `FragmentRepresentation`? If we still want to have that encapsulation, then having `NodeRepresentation` inside each `FragmentRepresentation` makes more sense to me.

**Comment by martint** [comment]
> DistributedPlanRepresentation is only used by the JsonRenderer. Extending it to TextRenderer might not be applicable. Curious to know your thoughts on this.

We don't need to use it *yet* for the text renderer, but we should design it to make it possible. In the future, as a separate change, we should adapt the text renderer.

**Comment by sumeethkc** [comment]
> > DistributedPlanRepresentation is only used by the JsonRenderer. Extending it to TextRenderer might not be applicable. Curious to know your thoughts on this.
> 
> We don't need to use it _yet_ for the text renderer, but we should design it to make it possible. In the future, as a separate change, we should adapt the text renderer.

I made the change to use the `DistributedPlanRepresentation` in `TextRenderer` in the same PR. Please let me know if this needs to be in a different PR.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java [hunk 17]
```diff
@@ -0,0 +1,495 @@
...
+import static java.util.stream.Collectors.toList;
+
+public class DistributedFragmentRepresentation
+{
+    private final String name;
+    private final String identifier;
+    private final List<TypedSymbol> outputs;
+    private final String details;
+    private final DistributedFragmentStats distributedFragmentStats;
+    private final Optional<PlanNodeStatsAndCostSummary> planNodeStatsAndCostSummary;
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java:49]: Are these the stats associated with `nodeRepresentation` below?

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java [hunk 18]
```diff
@@ -0,0 +1,495 @@
...
+    private final List<EstimateStats> estimateStats;
+    private final List<DistributedFragmentRepresentation> children;
+    private final Optional<WindowOperatorStatsRepresentation> windowOperatorStats;
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java:52]: Why is this here? It doesn't seem to belong to the fragment.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java [hunk 19]
```diff
@@ -0,0 +1,495 @@
...
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class DistributedFragmentRepresentation
+{
+    private final String name;
+    private final String identifier;
+    private final List<TypedSymbol> outputs;
+    private final String details;
+    private final DistributedFragmentStats distributedFragmentStats;
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java:48]: Call this `stats`

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java [hunk 20]
```diff
@@ -0,0 +1,495 @@
...
+    private final Optional<PlanNodeStatsAndCostSummary> planNodeStatsAndCostSummary;
+    private final List<EstimateStats> estimateStats;
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java:50]: Call this `estimates`

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java [hunk 21]
```diff
@@ -0,0 +1,495 @@
...
+        this.name = requireNonNull(name, "name is null");
+        this.identifier = requireNonNull(identifier, "identifier is null");
+        this.outputs = requireNonNull(outputs, "outputs is null");
+        this.details = requireNonNull(details, "details is null");
+        this.planNodeStatsAndCostSummary = planNodeStatsAndCostSummary;
+        this.estimateStats = estimateStats;
+        this.distributedFragmentStats = distributedFragmentStats;
+        this.children = children;
+        this.windowOperatorStats = windowOperatorStats;
+        this.nodeRepresentation = nodeRepresentation;
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java:77]: Add null checks

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java [hunk 22]
```diff
@@ -0,0 +1,495 @@
...
+        return windowOperatorStats;
+    }
+
+    @JsonIgnore
+    public NodeRepresentation getNodeRepresentation()
+    {
+        return nodeRepresentation;
+    }
+
+    public static DistributedFragmentRepresentation getPlanRepresentation(PlanRepresentation plan, NodeRepresentation node, Boolean verbose, Integer level)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java:140]: Some comments about this method:
* This method is misnamed. It returns a fragment representation, not a plan representation.
* What's the relationship between the `PlanRepresentation plan` argument and the `NodeRepresentation node` argument
* Use primitive types for `verbose` and `level`, instead of `Boolean` and `Integer`

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java [hunk 23]
```diff
@@ -0,0 +1,495 @@
...
+                    formatAsDataSize(cost.getMaxMemory()),
+                    formatAsDataSize(cost.getNetworkCost())));
+        }
+
+        return estimateStats;
+    }
+
+    private static Optional<PlanNodeStatsAndCostSummary> getReorderJoinStatsAndCost(NodeRepresentation node, boolean verbose)
+    {
+        if (verbose && node.getReorderJoinStatsAndCost().isPresent()) {
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java:206]: `verbose` should be a display-time concern. The representation object should contain all the relevant data, and the logic that prints the plan should choose what to include or exclude based on the verbose flag.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java [hunk 24]
```diff
@@ -217,6 +217,11 @@ private String toJson()
         return new JsonRenderer().render(representation);
     }
 
+    public PlanRepresentation getPlanRepresentation()
+    {
+        return representation;
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java:222]: This shouldn't be be exposed outside this class. It's an internal detail necessary for printing the logical plan. In fact, its initialization should be moved from the constructor to where it's actually used. The distributed plan representation should be handled in a similar manner. Specifically, the distributed plan should not be derived from the PlanRepresentation, but from the SubPlan and other structures.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java (hunk 23 above)
Comment (sumeethkc) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java:206]: Since I use JsonCodec, depending on whether `verbose` is present or not, I am creating the `windowOperatorStats` object. The JsonRenderer will just take the object and render using JsonCodec. Determining whether to render a member of class depending on whether `verbose` is present or not will make rendering more complex. I am not sure how to even do that actually with JsonCodec.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java (hunk 17 above)
Comment (sumeethkc) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java:49]: Yes, these are from `NodeRepresentation` if the `PlanNodeStats` is of type `WindoPlanNodeStats`. Hence, declared these as Optional.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java (hunk 18 above)
Comment (sumeethkc) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java:52]: Yes, these are from `NodeRepresentation` if the `PlanNodeStats` is of type `WindoPlanNodeStats`. Hence, declared these as Optional. But to render them using JsonCodec, I need to put them in the same class.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java (hunk 24 above)
Comment (sumeethkc) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/PlanPrinter.java:222]: Made that private. 

The PlanRepresentation contains the PlanNode. I would need to call `planRoot.accept()` after creating the PlanRepresentation so that the planRepresentation contains the node representation of all children. That is all done in the constructor.

For building the distributed plan from SubPlan, the PlanFragments' jsonRepresentation is built using PlanRepresentation in `PlanFragmenter::buildFragment`. So, I used the same method when building the DistributedpPlan representation

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java (hunk 22 above)
Comment (sumeethkc) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/DistributedFragmentRepresentation.java:140]: Renamed the method to `getFragmentRepresentation` and used primitive types.

PlanRepresentation is used to get the parent->child relationship between the nodes that are NodeRepresentation of all children for a given PlanNode. If NodeRepresentation contained children NodeRepresentations then PlanRepresentation wouldn't be needed but NodeRepresentation contains only PlanNodeIds as childrens. So, PlanRepresentation is used to get their NodeRepresentations.

**Comment by raunaqmorarka** [comment]
fyi https://github.com/trinodb/trino/pull/12694 implements `EXPLAIN (TYPE LOGICAL, FORMAT JSON)`. 
`EXPLAIN (TYPE DISTRIBUTED, FORMAT JSON)` and `EXPLAIN ANALYZE (TYPE DISTRIBUTED, FORMAT JSON)` are still left unimplemented though.

**Comment by mosabua** [comment]
:wave: @sumeethkc - a this stage `EXPLAIN (TYPE DISTRIBUTED, FORMAT JSON)` is also implemented and documented. As such I am closing this PR.
